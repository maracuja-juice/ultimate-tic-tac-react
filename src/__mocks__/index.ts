import {
  unfinishedBoardMock,
  movesForUnfinishedBoardMock,
} from './unfinishedBoardMock';
import {
  crossFinishedGameMock,
  circleFinishedGameMock,
} from './finishedGameDataMock';
import {
  circleFinishedBoardMock,
  crossFinishedBoardMock,
  movesForCircleFinishedBoardMock,
  movesForCrossFinishedBoardMock,
} from './finishedBoardMock';
import {
  activeBoardsForBoardWithTwoMovesMock,
  boardWithTwoMovesMock,
} from './boardWithTwoMovesMock';
import {
  boardWithThreeMovesMock,
  activeBoardsForBoardWithThreeMovesMock,
  movesForBoardWithThreeMovesMock,
} from './boardWithThreeMovesMock';
import {
  boardWithOneMoveMock,
  activeBoardsForBoardWithOneMoveMock,
} from './boardWithOneMoveMock';

export {
  unfinishedBoardMock,
  movesForUnfinishedBoardMock,
  crossFinishedGameMock,
  circleFinishedGameMock,
  circleFinishedBoardMock,
  movesForCircleFinishedBoardMock,
  crossFinishedBoardMock,
  movesForCrossFinishedBoardMock,
  boardWithTwoMovesMock,
  activeBoardsForBoardWithTwoMovesMock,
  boardWithThreeMovesMock,
  movesForBoardWithThreeMovesMock,
  activeBoardsForBoardWithThreeMovesMock,
  boardWithOneMoveMock,
  activeBoardsForBoardWithOneMoveMock,
};
