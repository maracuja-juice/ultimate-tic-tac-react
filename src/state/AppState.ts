import { Point } from '../util';

export interface AppState {
  currentGame: GameInformation;
  finishedGames: FinishedGameState[];
  analysisGame: AnalysisGame;
}

export interface GameInformation {
  board: SmallBoardInformation[];
  game: GameState;
  moves: MoveState[];
  activeBoards: Point[];
}

export interface FinishedGameState {
  id?: string;
  date: string;
  winner: Winner;
  gameState: SmallBoardInformation[];
  moves: MoveState[];
  saveState: string;
  errorMessage: string;
}

export interface AnalysisGame {
  id: string;
  board: SmallBoardInformation[];
  game: GameState;
  moves: MoveState[];
  activeBoards: Point[];
  currentMove: number;
}

export interface GameState {
  currentPlayer: Player;
  isFinished: boolean;
  winningPlayer: Winner;
}

export interface TileInformation {
  position: Point;
  value: TileValue;
}

export interface SmallBoardInformation extends TileInformation {
  tiles: SmallTileInformation[];
}

export interface SmallTileInformation extends TileInformation {
  boardPosition: Point;
}

export enum TileValue {
  Cross = 0,
  Circle = 1,
  Empty = 2,
  Destroyed = 3,
}

export enum Winner {
  Cross = 0,
  Circle = 1,
  None = 2,
  Draw = 3,
}

export enum Player {
  Cross = 0,
  Circle = 1,
}

export interface Move {
  boardPosition: Point;
  tilePosition: Point;
}

export interface MoveState extends Move {
  player: Player;
  moveNumber: number;
}

export interface GenericAction {
  type: string;
  // tslint:disable-next-line: no-any
  payload?: any;
}

export interface MarkSpecially {
  condition: boolean;
  position?: { boardPosition: Point; tilePosition: Point };
}
