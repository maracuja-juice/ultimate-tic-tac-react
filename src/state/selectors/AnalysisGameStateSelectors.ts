import { AnalysisGame, AppState, Player } from '../AppState';

export const getAnalysisGame = (
  state: AppState,
  id: string,
): AnalysisGame | undefined => {
  const finishedGame = state.finishedGames.find((g) => g.id === id);
  if (finishedGame) {
    return {
      id: finishedGame.id!,
      board: finishedGame.gameState,
      moves: finishedGame.moves,
      currentMove: finishedGame.moves.length,
      game: {
        currentPlayer: Player.Circle,
        winningPlayer: finishedGame.winner,
        isFinished: true,
      },
      activeBoards: [],
    };
  }

  return undefined;
};
