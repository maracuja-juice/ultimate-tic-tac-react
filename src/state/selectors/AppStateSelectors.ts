import { AnalysisGame, AppState, Player } from '../AppState';

export const getBoards = (state: AppState) => state.currentGame.board;

export const getCurrentPlayer = (state: AppState) =>
  state.currentGame.game.currentPlayer;

export const getWinningPlayer = (state: AppState) =>
  state.currentGame.game.winningPlayer;

export const getMoves = (state: AppState) => state.currentGame.moves;

export const getFinishedGame = (
  state: AppState,
  id: string,
): AnalysisGame | undefined => {
  const finishedGame = state.finishedGames.find((g) => g.id === id);
  if (finishedGame) {
    return {
      moves: finishedGame.moves,
      currentMove: finishedGame.moves.length,
      game: {
        currentPlayer: Player.Circle,
        winningPlayer: finishedGame.winner,
        isFinished: true,
      },
      activeBoards: [],
      board: finishedGame.gameState,
      id: finishedGame.id!,
    };
  }

  return undefined;
};
