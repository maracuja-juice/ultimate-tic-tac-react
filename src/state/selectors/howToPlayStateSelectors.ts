import { createSelector } from 'reselect';
import { getHowToPlay } from './appStateSelectors';
import { HowToPlay } from '../AppState';
import { howToPlaySteps } from '../howToPlay/howToPlaySteps';

export const getCurrentHowToPlayText = createSelector(
  [getHowToPlay],
  (howToPlay: HowToPlay) => {
    return howToPlaySteps[howToPlay.stepNumber].text;
  },
);
