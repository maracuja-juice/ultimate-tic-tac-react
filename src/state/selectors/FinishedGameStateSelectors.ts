import { createSelector } from 'reselect';
import { getBoards, getIsReplay, getMoves, getWinningPlayer } from './AppStateSelectors';

export const getWinningPlayerAsString = createSelector( [getWinningPlayer], (winningPlayer => {
    if (winningPlayer === 0) {
        return 'X';
    } else if (winningPlayer === 1) {
        return 'Y';
    } else if (winningPlayer === null) {
        return null;
    } else {
        return undefined;
    }
}) );

export const getFinishedGameData = createSelector(
    [getWinningPlayerAsString, getIsReplay, getBoards, getMoves],
    ( winningPlayer, isReplay, boards, moves ) => {
        return {
            winner: winningPlayer,
            gameState: boards,
            isReplay,
            moves
        };
    } );