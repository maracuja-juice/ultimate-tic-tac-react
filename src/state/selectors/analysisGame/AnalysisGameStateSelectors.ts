import { AnalysisGame, AppState, FinishedGameState } from '../../AppState';
import { createSelector } from 'reselect';
import { getFinishedGames } from '../AppStateSelectors';

const mapFinishedGameToAnalysisGame = (
  finishedGame: FinishedGameState,
): AnalysisGame => {
  return {
    id: finishedGame.id!,
    board: finishedGame.gameState,
    moves: finishedGame.moves,
    currentMove: finishedGame.moves.length,
    game: {
      currentPlayer: finishedGame.moves[finishedGame.moves.length - 2].player,
      winningPlayer: finishedGame.winner,
      isFinished: true,
    },
    activeBoards: [],
  };
};

export const getIdToFetch = (state: AppState, id: string) => id;

export const getAnalysisGameById = createSelector(
  [getFinishedGames, getIdToFetch],
  (finishedGames, id): AnalysisGame | undefined => {
    const finishedGame = finishedGames.find((g) => g.id === id);
    if (finishedGame) {
      return mapFinishedGameToAnalysisGame(finishedGame);
    }

    return undefined;
  },
);

export const getLatestAnalysisGame = createSelector(
  [getFinishedGames],
  (finishedGames: FinishedGameState[]): AnalysisGame | undefined => {
    if (finishedGames.length > 0) {
      const latestFinishedGame = finishedGames[finishedGames.length - 1];
      return mapFinishedGameToAnalysisGame(latestFinishedGame);
    }

    return undefined;
  },
);
