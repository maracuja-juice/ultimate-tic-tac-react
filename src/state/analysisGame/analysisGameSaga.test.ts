import { expectSaga } from 'redux-saga-test-plan';
import { select } from 'redux-saga/effects';
import { Player, Winner } from '../AppState';
import loadFinishedGameSaga from './analysisGameSaga';
import { getAnalysisGameById } from '../selectors/AnalysisGameStateSelectors';
import { crossFinishedBoardMock } from '../../__mocks__';
import {
  LOAD_FINISHED_GAME_BY_ID,
  SET_ANALYSIS_GAME,
} from './analysisGameActions';

describe('loadFinishedGameSaga', () => {
  const analysisGame = {
    id: '1',
    board: crossFinishedBoardMock,
    activeBoards: [],
    moves: [
      {
        moveNumber: 1,
        tilePosition: { x: 0, y: 0 },
        boardPosition: { x: 0, y: 0 },
        player: Player.Cross,
      },
    ],
    currentMove: 1,
    game: {
      currentPlayer: Player.Circle,
      winningPlayer: Winner.Cross,
      isFinished: true,
    },
  };

  it('should dispatch the setAnalysisGame action for the analysisGame with the given id', () => {
    return expectSaga(loadFinishedGameSaga)
      .provide([[select(getAnalysisGameById, '1'), analysisGame]])
      .put({ type: SET_ANALYSIS_GAME, payload: analysisGame })
      .dispatch({
        type: LOAD_FINISHED_GAME_BY_ID,
        payload: '1',
      })
      .silentRun();
  });

  // if more put effects happen: this catches it + this checks for the order
  it('should match snapshot', () => {
    return expectSaga(loadFinishedGameSaga)
      .provide([[select(getAnalysisGameById, '1'), analysisGame]])
      .dispatch({
        type: LOAD_FINISHED_GAME_BY_ID,
        payload: '1',
      })
      .silentRun()
      .then((result) => {
        expect(result.toJSON()).toMatchSnapshot();
      });
  });
});
