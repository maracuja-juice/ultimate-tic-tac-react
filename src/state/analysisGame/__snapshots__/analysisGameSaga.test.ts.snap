// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loadFinishedGameSaga should match snapshot 1`] = `
Object {
  "fork": Array [
    Object {
      "@@redux-saga/IO": true,
      "FORK": Object {
        "args": Array [
          "analysisGameReducer/load-finished-game-by-id",
          "@@redux-saga-test-plan/json/function/loadFinishedGameById",
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/takeEvery",
      },
    },
    Object {
      "@@redux-saga/IO": true,
      "FORK": Object {
        "args": Array [
          "analysisGameReducer/load-latest-finished-game",
          "@@redux-saga-test-plan/json/function/loadLatestFinishedGame",
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/takeEvery",
      },
    },
    Object {
      "@@redux-saga/IO": true,
      "FORK": Object {
        "args": Array [
          Object {
            "payload": "1",
            "type": "analysisGameReducer/load-finished-game-by-id",
          },
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/<anonymous>",
      },
    },
  ],
  "put": Array [
    Object {
      "@@redux-saga/IO": true,
      "PUT": Object {
        "action": Object {
          "payload": Object {
            "activeBoards": Array [],
            "board": Array [
              Object {
                "position": Object {
                  "x": 0,
                  "y": 0,
                },
                "tiles": Array [
                  Object {
                    "boardPosition": Object {
                      "x": 0,
                      "y": 0,
                    },
                    "position": Object {
                      "x": 0,
                      "y": 0,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 0,
                      "y": 0,
                    },
                    "position": Object {
                      "x": 0,
                      "y": 1,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 0,
                      "y": 0,
                    },
                    "position": Object {
                      "x": 0,
                      "y": 2,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 0,
                      "y": 0,
                    },
                    "position": Object {
                      "x": 1,
                      "y": 0,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 0,
                      "y": 0,
                    },
                    "position": Object {
                      "x": 1,
                      "y": 1,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 0,
                      "y": 0,
                    },
                    "position": Object {
                      "x": 1,
                      "y": 2,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 0,
                      "y": 0,
                    },
                    "position": Object {
                      "x": 2,
                      "y": 0,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 0,
                      "y": 0,
                    },
                    "position": Object {
                      "x": 2,
                      "y": 1,
                    },
                    "value": 0,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 0,
                      "y": 0,
                    },
                    "position": Object {
                      "x": 2,
                      "y": 2,
                    },
                    "value": 2,
                  },
                ],
                "value": 2,
              },
              Object {
                "position": Object {
                  "x": 0,
                  "y": 1,
                },
                "tiles": Array [
                  Object {
                    "boardPosition": Object {
                      "x": 0,
                      "y": 1,
                    },
                    "position": Object {
                      "x": 0,
                      "y": 0,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 0,
                      "y": 1,
                    },
                    "position": Object {
                      "x": 0,
                      "y": 1,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 0,
                      "y": 1,
                    },
                    "position": Object {
                      "x": 0,
                      "y": 2,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 0,
                      "y": 1,
                    },
                    "position": Object {
                      "x": 1,
                      "y": 0,
                    },
                    "value": 1,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 0,
                      "y": 1,
                    },
                    "position": Object {
                      "x": 1,
                      "y": 1,
                    },
                    "value": 1,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 0,
                      "y": 1,
                    },
                    "position": Object {
                      "x": 1,
                      "y": 2,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 0,
                      "y": 1,
                    },
                    "position": Object {
                      "x": 2,
                      "y": 0,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 0,
                      "y": 1,
                    },
                    "position": Object {
                      "x": 2,
                      "y": 1,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 0,
                      "y": 1,
                    },
                    "position": Object {
                      "x": 2,
                      "y": 2,
                    },
                    "value": 2,
                  },
                ],
                "value": 2,
              },
              Object {
                "position": Object {
                  "x": 0,
                  "y": 2,
                },
                "tiles": Array [
                  Object {
                    "boardPosition": Object {
                      "x": 0,
                      "y": 2,
                    },
                    "position": Object {
                      "x": 0,
                      "y": 0,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 0,
                      "y": 2,
                    },
                    "position": Object {
                      "x": 0,
                      "y": 1,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 0,
                      "y": 2,
                    },
                    "position": Object {
                      "x": 0,
                      "y": 2,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 0,
                      "y": 2,
                    },
                    "position": Object {
                      "x": 1,
                      "y": 0,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 0,
                      "y": 2,
                    },
                    "position": Object {
                      "x": 1,
                      "y": 1,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 0,
                      "y": 2,
                    },
                    "position": Object {
                      "x": 1,
                      "y": 2,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 0,
                      "y": 2,
                    },
                    "position": Object {
                      "x": 2,
                      "y": 0,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 0,
                      "y": 2,
                    },
                    "position": Object {
                      "x": 2,
                      "y": 1,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 0,
                      "y": 2,
                    },
                    "position": Object {
                      "x": 2,
                      "y": 2,
                    },
                    "value": 2,
                  },
                ],
                "value": 2,
              },
              Object {
                "position": Object {
                  "x": 1,
                  "y": 0,
                },
                "tiles": Array [
                  Object {
                    "boardPosition": Object {
                      "x": 1,
                      "y": 0,
                    },
                    "position": Object {
                      "x": 0,
                      "y": 0,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 1,
                      "y": 0,
                    },
                    "position": Object {
                      "x": 0,
                      "y": 1,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 1,
                      "y": 0,
                    },
                    "position": Object {
                      "x": 0,
                      "y": 2,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 1,
                      "y": 0,
                    },
                    "position": Object {
                      "x": 1,
                      "y": 0,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 1,
                      "y": 0,
                    },
                    "position": Object {
                      "x": 1,
                      "y": 1,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 1,
                      "y": 0,
                    },
                    "position": Object {
                      "x": 1,
                      "y": 2,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 1,
                      "y": 0,
                    },
                    "position": Object {
                      "x": 2,
                      "y": 0,
                    },
                    "value": 0,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 1,
                      "y": 0,
                    },
                    "position": Object {
                      "x": 2,
                      "y": 1,
                    },
                    "value": 0,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 1,
                      "y": 0,
                    },
                    "position": Object {
                      "x": 2,
                      "y": 2,
                    },
                    "value": 0,
                  },
                ],
                "value": 0,
              },
              Object {
                "position": Object {
                  "x": 1,
                  "y": 1,
                },
                "tiles": Array [
                  Object {
                    "boardPosition": Object {
                      "x": 1,
                      "y": 1,
                    },
                    "position": Object {
                      "x": 0,
                      "y": 0,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 1,
                      "y": 1,
                    },
                    "position": Object {
                      "x": 0,
                      "y": 1,
                    },
                    "value": 0,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 1,
                      "y": 1,
                    },
                    "position": Object {
                      "x": 0,
                      "y": 2,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 1,
                      "y": 1,
                    },
                    "position": Object {
                      "x": 1,
                      "y": 0,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 1,
                      "y": 1,
                    },
                    "position": Object {
                      "x": 1,
                      "y": 1,
                    },
                    "value": 0,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 1,
                      "y": 1,
                    },
                    "position": Object {
                      "x": 1,
                      "y": 2,
                    },
                    "value": 1,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 1,
                      "y": 1,
                    },
                    "position": Object {
                      "x": 2,
                      "y": 0,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 1,
                      "y": 1,
                    },
                    "position": Object {
                      "x": 2,
                      "y": 1,
                    },
                    "value": 0,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 1,
                      "y": 1,
                    },
                    "position": Object {
                      "x": 2,
                      "y": 2,
                    },
                    "value": 1,
                  },
                ],
                "value": 0,
              },
              Object {
                "position": Object {
                  "x": 1,
                  "y": 2,
                },
                "tiles": Array [
                  Object {
                    "boardPosition": Object {
                      "x": 1,
                      "y": 2,
                    },
                    "position": Object {
                      "x": 0,
                      "y": 0,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 1,
                      "y": 2,
                    },
                    "position": Object {
                      "x": 0,
                      "y": 1,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 1,
                      "y": 2,
                    },
                    "position": Object {
                      "x": 0,
                      "y": 2,
                    },
                    "value": 0,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 1,
                      "y": 2,
                    },
                    "position": Object {
                      "x": 1,
                      "y": 0,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 1,
                      "y": 2,
                    },
                    "position": Object {
                      "x": 1,
                      "y": 1,
                    },
                    "value": 1,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 1,
                      "y": 2,
                    },
                    "position": Object {
                      "x": 1,
                      "y": 2,
                    },
                    "value": 0,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 1,
                      "y": 2,
                    },
                    "position": Object {
                      "x": 2,
                      "y": 0,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 1,
                      "y": 2,
                    },
                    "position": Object {
                      "x": 2,
                      "y": 1,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 1,
                      "y": 2,
                    },
                    "position": Object {
                      "x": 2,
                      "y": 2,
                    },
                    "value": 0,
                  },
                ],
                "value": 0,
              },
              Object {
                "position": Object {
                  "x": 2,
                  "y": 0,
                },
                "tiles": Array [
                  Object {
                    "boardPosition": Object {
                      "x": 2,
                      "y": 0,
                    },
                    "position": Object {
                      "x": 0,
                      "y": 0,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 2,
                      "y": 0,
                    },
                    "position": Object {
                      "x": 0,
                      "y": 1,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 2,
                      "y": 0,
                    },
                    "position": Object {
                      "x": 0,
                      "y": 2,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 2,
                      "y": 0,
                    },
                    "position": Object {
                      "x": 1,
                      "y": 0,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 2,
                      "y": 0,
                    },
                    "position": Object {
                      "x": 1,
                      "y": 1,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 2,
                      "y": 0,
                    },
                    "position": Object {
                      "x": 1,
                      "y": 2,
                    },
                    "value": 1,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 2,
                      "y": 0,
                    },
                    "position": Object {
                      "x": 2,
                      "y": 0,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 2,
                      "y": 0,
                    },
                    "position": Object {
                      "x": 2,
                      "y": 1,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 2,
                      "y": 0,
                    },
                    "position": Object {
                      "x": 2,
                      "y": 2,
                    },
                    "value": 0,
                  },
                ],
                "value": 2,
              },
              Object {
                "position": Object {
                  "x": 2,
                  "y": 1,
                },
                "tiles": Array [
                  Object {
                    "boardPosition": Object {
                      "x": 2,
                      "y": 1,
                    },
                    "position": Object {
                      "x": 0,
                      "y": 0,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 2,
                      "y": 1,
                    },
                    "position": Object {
                      "x": 0,
                      "y": 1,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 2,
                      "y": 1,
                    },
                    "position": Object {
                      "x": 0,
                      "y": 2,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 2,
                      "y": 1,
                    },
                    "position": Object {
                      "x": 1,
                      "y": 0,
                    },
                    "value": 1,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 2,
                      "y": 1,
                    },
                    "position": Object {
                      "x": 1,
                      "y": 1,
                    },
                    "value": 1,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 2,
                      "y": 1,
                    },
                    "position": Object {
                      "x": 1,
                      "y": 2,
                    },
                    "value": 1,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 2,
                      "y": 1,
                    },
                    "position": Object {
                      "x": 2,
                      "y": 0,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 2,
                      "y": 1,
                    },
                    "position": Object {
                      "x": 2,
                      "y": 1,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 2,
                      "y": 1,
                    },
                    "position": Object {
                      "x": 2,
                      "y": 2,
                    },
                    "value": 2,
                  },
                ],
                "value": 1,
              },
              Object {
                "position": Object {
                  "x": 2,
                  "y": 2,
                },
                "tiles": Array [
                  Object {
                    "boardPosition": Object {
                      "x": 2,
                      "y": 2,
                    },
                    "position": Object {
                      "x": 0,
                      "y": 0,
                    },
                    "value": 1,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 2,
                      "y": 2,
                    },
                    "position": Object {
                      "x": 0,
                      "y": 1,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 2,
                      "y": 2,
                    },
                    "position": Object {
                      "x": 0,
                      "y": 2,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 2,
                      "y": 2,
                    },
                    "position": Object {
                      "x": 1,
                      "y": 0,
                    },
                    "value": 1,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 2,
                      "y": 2,
                    },
                    "position": Object {
                      "x": 1,
                      "y": 1,
                    },
                    "value": 0,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 2,
                      "y": 2,
                    },
                    "position": Object {
                      "x": 1,
                      "y": 2,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 2,
                      "y": 2,
                    },
                    "position": Object {
                      "x": 2,
                      "y": 0,
                    },
                    "value": 1,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 2,
                      "y": 2,
                    },
                    "position": Object {
                      "x": 2,
                      "y": 1,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 2,
                      "y": 2,
                    },
                    "position": Object {
                      "x": 2,
                      "y": 2,
                    },
                    "value": 0,
                  },
                ],
                "value": 1,
              },
            ],
            "currentMove": 1,
            "game": Object {
              "currentPlayer": 1,
              "isFinished": true,
              "winningPlayer": 0,
            },
            "id": "1",
            "moves": Array [
              Object {
                "boardPosition": Object {
                  "x": 0,
                  "y": 0,
                },
                "moveNumber": 1,
                "player": 0,
                "tilePosition": Object {
                  "x": 0,
                  "y": 0,
                },
              },
            ],
          },
          "type": "analysisGameReducer/set-analysis-game",
        },
        "channel": null,
      },
    },
  ],
  "select": Array [
    Object {
      "@@redux-saga/IO": true,
      "SELECT": Object {
        "args": Array [
          "1",
        ],
        "selector": "@@redux-saga-test-plan/json/function/<anonymous>",
      },
    },
  ],
  "take": Array [
    Object {
      "@@redux-saga/IO": true,
      "TAKE": Object {
        "pattern": "analysisGameReducer/load-finished-game-by-id",
      },
    },
    Object {
      "@@redux-saga/IO": true,
      "TAKE": Object {
        "pattern": "analysisGameReducer/load-latest-finished-game",
      },
    },
    Object {
      "@@redux-saga/IO": true,
      "TAKE": Object {
        "pattern": "analysisGameReducer/load-finished-game-by-id",
      },
    },
  ],
}
`;
